File: src/main/scala/2017/day-13.scala
package adventofcode2017

import scala.io.Source

object Day13 {

  def isCoughtOnLayer(layer: Option[Int], step: Int): Boolean = {
    layer match {
      case None => false
      case Some(depth) => 
        val devider = depth*2 - 2
        step % devider == 0
    }
  }

  def willGetCought(layers: Int, map: Map[Int, Int], delay: Int): Boolean = {
    (0 until layers+1).foldLeft(false) {
      (cought, step) => cought || 
        isCoughtOnLayer(map.get(step), step+delay)
    }
  }

  def findDelay(layers: Int, map: Map[Int, Int], delay: Int = 0): Int = {
    if (!willGetCought(layers, map, delay)) delay
    else findDelay(layers, map, delay+1)
  }

  lazy val input: String = Source.fromFile(s"2017/input/day-13.input").mkString.trim
   def main(args: Array[String]): Unit = {
    val map = input.lines.foldLeft(Map[Int, Int]()) {
      (map, row) => 
        val layer::depth::_ = row.split(": ").toList
        map + (layer.toInt -> depth.toInt)
    }
    val layers = map.keys.foldLeft(0) { Math.max(_, _) }

    val sev = (0 until layers+1).foldLeft(0) {
      (sev, step) => 
        if (isCoughtOnLayer(map.get(step), step)) 
          sev + (map.getOrElse(step, 0) * step)
        else
          sev
    }
    println(s"part1 $sev")

    val delay = findDelay(layers, map)
    println(s"part2 $delay")
   }
}

File: 2017/js/day-13.js
const fs = require('fs');
const input = fs.readFileSync('2017/input/day-13.input').toString();

const map = input.split('\n').reduce((map, e) => {
  const values = e.split(': ');
  map[values[0]] = +values[1];
  return map;
}, {});

const layers = Math.max(...Object.keys(map).map(x => +x));

function isCoughtOnLayer(layer, step) {
  const depth = layer;
  if (layer !== void 0) {
    const devider = depth + depth - 2;
    return step % devider === 0;
  }
  return false;
}

const coughts = [];
for(let step=0; step<=layers; step++) {
  if (isCoughtOnLayer(map[step], step)) {
    coughts.push(step);
  }
}
const sev = coughts.reduce((res, x) => res + x * map[x], 0);
console.log('part1', sev);

function willGetCought(delay) {
  for(let step=0; step<=layers; step++) {
    if (isCoughtOnLayer(map[step], step + delay)) {
      return true;
    }
  }
  return false;
}

let delay = -1;
let pass = false;
while (!pass && delay <= 10000000) {
  delay++;
  pass = !willGetCought(delay);
}
console.log('part2', delay);